### ISSUE TITLE
Restart a Cancelled Upload

### ISSUE THREAD

**Issue by ar27111994** [bleenco/ngx-uploader#453]
Hi, how can I restart a cancelled file from queue? I have tried the `uploadFile` event, but it has two problems .

1. The event has to be emitted twice for the upload to start again.

2. The `cancel` even stops working after this.

Therefore, please help.

**Comment by ar27111994** [comment]
Anybody home? Is this plugin actively maintained?

**Comment by jkuri** [comment]
hi @ar27111994 and sorry for super late response. It is maintained but currently I am very busy with other stuff.
The problem you described is a known issue and I will try to fix that till the end of this week.

**Comment by ar27111994** [comment]
Thanks. And I've one more request, can you add max file size option in upload options interface? Currently, its not supported at all, and if I am not wrong, is scheduled for next release.

**Comment by jkuri** [comment]
For the max file size option you can filter upload files yourself, before starting upload, ie;

```ts
sizeLimit = 1024 * 10;
uploadInput = new EventEmitter<UploadInput>();

onUploadOutput(output: UploadOutput): void {
  if (output.type === 'addedToQueue') {
    const file: UploadFile = output.file;
    if (file.size > this.sizeLimit) {
      const ev: UploadInput = { type: 'remove', id: file.id };
      this.uploadInput.emit(ev);
    }    
  }
}
```

Please note that I did not test that code, please let me know if it works. Also, I agree it would be more convenient to have this feature implemented inside library althought it is achiveable currently as well.

**Comment by ar27111994** [comment]
Hi @jkuri no the `remove` even doesn't work, it still queues it up and uploads it. Here's my current implementation:
![ngx-uploader max-size](https://user-images.githubusercontent.com/13036687/43820851-cf397466-9b00-11e8-84cd-7485cc8ccbc6.JPG)
![ngx-uploader max-size 2](https://user-images.githubusercontent.com/13036687/43820850-cf034148-9b00-11e8-8979-a79c9eb70b6b.JPG)

And it just hides the rejected files from ui, but still queues them up and uploads them behind the scenes. Hence I don't think there's any other way of validating and restricting file size except for it to be in upload options and checked before queuing up files.

**Comment by ar27111994** [comment]
Hi @jkuri , any update on this?

**Comment by ar27111994** [comment]
guys should we switch to another uploader? :/ I can create a pr for fixing max size issue but I really don't understand the problem with re-uploading a cancelled file.

**Comment by Sabdeth** [comment]
I am having the same problem. I am testing the **cancelAll** feature and I am unable to restart the upload process after cancelling. It sets the file status to 3 (Cancelled) for all files whose status == Uploading  | | Done

I then tried using **cancel** event. So that it only cancels the file whose file status == Uploading. See below code snipet.
```
cancelUpload(id: string): void {
   this.uploadInput.emit({ type: 'cancel', id: id });
 }
```


```
cancelAllUpload(): void {
   this.files.filter(s=>s.progress.status === UploadStatus.Uploading).forEach(f=>this.cancelUpload(f.id));
}
```
It calls cancelUpload function on each files whose status == Uploading. 
But when I try to resume my upload nothing happens even for the files who status == Queue.

Is there a workaround for this issue ?
Basically I would like to resume uploading of files where status == Queue | | status == Cancelled . 

Thanks!

**Comment by JianhuisHuang** [comment]
excuse me,I can't find the solution for this bug "restarting cancel upload",can you help me ?
thanks @jkuri @ar27111994

**Comment by victor-panaite** [comment]
A solution that worked for me in case of restarting the upload was to add a reference on the component 
```
<input
      #fakeUpload
      type="file"
      ngFileSelect
      [options]="options"
      [uploadInput]="uploadInput"
      (uploadOutput)="onUploadOutput($event)"
    />
```

and then right before creating the event for `this.uploadInput.emit(event);` I've manually trigger a change event on that reference.

```
const ev = new Event('change');    
this.fakeUploadInput.nativeElement.dispatchEvent(ev);
const event: UploadInput = {
      file: this.file,
      headers: {
        Authorization: `JWT ${token}`
      },
      method: 'POST',
      type: 'uploadAll',
      url: this.url
    };
this.uploadInput.emit(event);
```

**Comment by antrikshpatel** [comment]
> A solution that worked for me in case of restarting the upload was to add a reference on the component
> 
> ```
> <input
>       #fakeUpload
>       type="file"
>       ngFileSelect
>       [options]="options"
>       [uploadInput]="uploadInput"
>       (uploadOutput)="onUploadOutput($event)"
>     />
> ```
> 
> and then right before creating the event for `this.uploadInput.emit(event);` I've manually trigger a change event on that reference.
> 
> ```
> const ev = new Event('change');    
> this.fakeUploadInput.nativeElement.dispatchEvent(ev);
> const event: UploadInput = {
>       file: this.file,
>       headers: {
>         Authorization: `JWT ${token}`
>       },
>       method: 'POST',
>       type: 'uploadAll',
>       url: this.url
>     };
> this.uploadInput.emit(event);
> ```

I tried your suggestion. The problem i faced is it adds duplicate files to the list.
